marshal/marshal.go

```diff
@@ -28,91 +30,371 @@ const (
+
+	// Lock protects the following members; you must have this lock in order to
+	// read from or write to these.
+	lock      sync.RWMutex
+	consumers []*Consumer
+}
+
 // NewMarshaler connects to a cluster (given broker addresses) and prepares to handle marshalling
 // requests. Given the way this system works, the marshaler has to process all messages in the
 // topic before it's safely able to begin operating. This might take a while.
 func NewMarshaler(clientID, groupID string, brokers []string) (*Marshaler, error) {
-	// TODO: It might be nice to make the marshaler agnostic of clients and able to support
-	// requests from N clients/groups. For now, though, we require instantiating a new
-	// marshaler for every client/group.
-	brokerConf := kafka.NewBrokerConf("PortalMarshal")
-	broker, err := kafka.Dial(brokers, brokerConf)
+	cluster, err := getOrCreateCluster(brokers)
 	if err != nil {
 		return nil, err
 	}
 
 	// Get offset coordinator so we can look up (and save) committed offsets later.
-	coordinatorConf := kafka.NewOffsetCoordinatorConf(groupID)
-	coordinator, err := broker.OffsetCoordinator(coordinatorConf)
+	coordinator, err := cluster.getOffsetCoordinator(groupID)
 	if err != nil {
 		return nil, err
 	}
 
 	m := &Marshaler{
 		quit:       new(int32),
-		rsteps:     new(int32),
+		cluster:    cluster,
 		instanceID: newInstanceID(),
 		clientID:   clientID,
 		groupID:    groupID,
-		kafka:      broker,
 		offsets:    coordinator,
-		producer:   broker.Producer(kafka.NewProducerConf()),
-		topics:     make(map[string]int),
-		groups:     make(map[string]map[string]*topicState),
-		jitters:    make(chan time.Duration, 100),
 	}
 
-	// Do an initial metadata fetch, this will block a bit
-	err = m.refreshMetadata()
-	if err != nil {
-		return nil, fmt.Errorf("Failed to get metadata: %s", err)
+	return m, nil
+}
+
+// newInstanceID creates a new random instance ID for use inside Marshal messages. This
+// is generated new every time we restart.
+func newInstanceID() string {
+	// A UUID4 starts with 8 random characters, so let's use that as our instance ID.
+	// This should be a good tradeoff between randomness and brevity.
+	return uuid.New()[0:8]
+}
+
+// addNewConsumer is called when a new Consumer is created. This allows Marshal to keep
+// track of the consumers that exist so we can operate on them later if needed.
+func (m *Marshaler) addNewConsumer(c *Consumer) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.consumers = append(m.consumers, c)
+}
+
+// getClaimedPartitionState returns a topicState iff it is claimed by the current Marshaler.
+// Else, an error is returned. This is on the Marshaler becomes it's a helper to only return
+// a claim that is presently valid and owned by us.
+func (m *Marshaler) getClaimedPartitionState(topicName string, partID int) (
+	*topicState, error) {
+
+	// Get partition state of whatever happens to be here
+	topic := m.cluster.getPartitionState(m.groupID, topicName, partID)
```

Comment:
Added this to the cluster termination logic now. When you terminate a cluster, it terminates all marshalers which in turn terminate all consumers.